### PR TITLE
Consequently use /run instead of /var/run

### PR DESCRIPTION
**Pull request by MichaIng**
Companion PR to: https://github.com/pi-hole/pi-hole/pull/3248

Signed-off-by: MichaIng <micha@dietpi.com>

## Description
Consequently use `/run` instead of deprecated `/var/run` for runtime files, to satisfy FHS and systemd and avoid confusion because of conflicting paths, which are valid only because of `/var/run -> /run` symlink.

## Motivation and Context
See: https://github.com/pi-hole/pi-hole/pull/3248#issuecomment-609713592

## How Has This Been Tested?
It has not been tested, however the rationale behind it is assumed trivial enough. A test can be done, if wanted.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Satisfy code and setup standards and consistency, to pro-actively avoid possible future issues and reduce chance for confusion.

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly: https://github.com/pi-hole/docs/pull/313

### CODE REVIEW

File: docker-compose-jwilder-proxy.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ services:
     environment:
       DEFAULT_HOST: pihole.yourDomain.lan
     volumes:
-      - '/var/run/docker.sock:/tmp/docker.sock'
+      - '/run/docker.sock:/tmp/docker.sock'
```
Comment (dschaper) [inline docker-compose-jwilder-proxy.yml:13]: This will break things.

File: circle-test.sh [hunk 2]
```diff
@@ -15,7 +15,7 @@ fi
 docker build -t image_pipenv -f Dockerfile_build .
 env > /tmp/env
 docker run --rm \
-    -v /var/run/docker.sock:/var/run/docker.sock \
+    -v /run/docker.sock:/run/docker.sock \
```
Comment (dschaper) [inline circle-test.sh:18]: This will break things.

File: docker-compose-traefik-proxy.md [hunk 3]
```diff
@@ -30,7 +30,7 @@ services:
       - "443:443"
       - "8080:8080"
     volumes:
-      - /var/run/docker.sock:/var/run/docker.sock
+      - /run/docker.sock:/run/docker.sock
```
Comment (dschaper) [inline docker-compose-traefik-proxy.md:33]: This will break things.

**Comment by dschaper** [comment]
Please try this in a docker environment.

**Comment by dschaper** [comment]
>With -H it is possible to make the Docker daemon to listen on a specific IP and port. By default, it will listen on unix:///var/run/docker.sock to allow only local connections by the root user.

**Comment by MichaIng** [comment]
Thanks for your feedback.

To be true, I am no Docker specialist. However I pulled the current pihole/pihole:latest container and replaced all `/var/run` occurrence with `/run` in the files affected by this PR and the container boots up fine, Pi-hole web UI is accessible and fully functional. But the docker-compose and circle-test files are not present, guess they are somehow compiled in? 

Your concerns are about the volumes:
In case of `- /var/run/docker.sock:/var/run/docker.sock` the first defines the path on the host, on which the reason for doing all this changes is valid, so this can and should be changed to comply with modern standards:
```
2020-05-16 22:02:14 root@VM-Buster:~# mv /var/run/docker.sock /run/docker.sock
mv: '/var/run/docker.sock' and '/run/docker.sock' are the same file
2020-05-16 22:02:40 root@VM-Buster:~# readlink -f /var/run/docker.sock
/run/docker.sock
```
The second defines the bind mount point inside the container. There I am not sure if `/var/run` is a symlink to `/run` like on the host. But since the docker run command explicitly creates these volumes it should be possible to tell it (whatever uses this socket from inside the container) to use the `/run/docker.sock` as well?

The man page entry you linked is about the socket file on host side, so even that Docker still uses the deprecated location, as of the symlink it doesn't matter there.
___________
Btw how do I build the container from sources, to test it correctly (instead of manipulating overlay files from pulled downstream container)? The docker build command fails silently on `install.sh`, I guess due to `cat /etc/docker-pi-hole-version` which does not exist. Do I need to prepare the build environment somehow?

**Comment by dschaper** [comment]
@diginc It's all yours.

**Comment by diginc** [comment]
Do not touch either /var/run/docker.sock or the s6 files if you want this to be merged.  undo those changes.

**Comment by MichaIng** [comment]
@diginc 
Okay will do that.

As of my response above and the reason for this change, is there any specific reason why this must be in `/var/run`? As said on the host both are identical and within the container we should be able to control it.

**Comment by diginc** [comment]
Docker's installers put it in /var/run/, not /run/.  Moving the default location is more work and goes against the default creating odd-ball systems people will just be confused by.

dito for s6-overlay, google s6-overlay and you'll see all their docs say /var/run/s6.

**Comment by MichaIng** [comment]
@diginc 
As it is a symlink, the docker installers actually put it in `/run`. But yeah since Docker uses deprecated `/var/run` all across their docs and scripts and services, it is an argument to avoid confusion an leave it is now. Earlier or later Docker will adjust this, then we can follow.

**Comment by diginc** [comment]
Fair point, documentation for any sort of docker-in-docker is still saying mount /var/run/docker though. Haven't read anywhere saying docker users should be switching over to /run/ but curious if you have any official sources saying to.

- https://hub.docker.com/_/docker/
- https://docs.traefik.io/v1.4/user-guide/docker-and-lets-encrypt/

**Comment by MichaIng** [comment]
@diginc 
See here, sums up everything well: https://github.com/pi-hole/pi-hole/pull/3248#issuecomment-609713592

**Comment by diginc** [comment]
Thanks, I see now. Seems like the trickle down effect is taking some time for updates to reach all parts of the chain/pipeline:  OSs/OS Best Practices -> Packages -> Meta Packages & Package Consumers.

We're at the 3rd layer, consuming docker/s6 (second layer) who are still semi-behind on updating to advocate for / advertise the new preference for /run, them being behind was causing my hesitation.

Alpine OS is not symlinking /var/run to /run/ so that maybe part of why docker is lagging behind too, they're heavily invested in alpine containers. It is a bit of a pain to have to know 'is this container alpine or debian?' in order to know if you should be linking to /var/run or /run. It also depends on the docker.sock consuming source code, I'm sure official docker in docker probably looks at both locations but I wouldn't confidently state that for any 3rd party docker.socket utilizing apps, who may also be on alpine bases.

I can take a look at updating s6 at a later date, will need to go look and see if their github as any requests for them to update to /run/ too.

**Comment by MichaIng** [comment]
I checked the s6 repo itself, and it seems to not directly use or expect these locations at all:
- https://github.com/skarnet/s6/search?q=%2Fvar%2Frun&type=Code
- https://github.com/skarnet/s6/search?q=%2Frun&type=Code

This somehow makes sense as the supervisor basically has nothing to do with where services place their sockets, PID files and other runtime files. Seems to be up to the user if and where to place it.

`s6-overlay` for containers on the other hand uses `/var/run` throughout indeed: https://github.com/just-containers/s6-overlay/search?q=%2Frun&type=Code
Also I couldn't find any request to change this. Could be worth a request, although priority will clearly be small as it does not really fix anything. They could follow most distros and create the symlink for now. So devs can use `/run` as well as `/var/run` for their software, outside and inside the container, without risking to break something, so can choose whether to keep old or follow the new standard.

**Comment by MichaIng** [comment]
Is anything missing here? FTL is creating the socket file at `/run/pihole/FTL.sock` for a while now, so if that symlink did not exist in the container system, it would fail: https://github.com/pi-hole/FTL/commit/0fd1e341da3b0a637f9de9b105b815a4db3f8d83

**Comment by dschaper** [comment]
I'm not keen on making changes for the sake of making changes.

**Comment by MichaIng** [comment]
Please read through related issues which contains the history and reasons why all other Pi-hole team members agreed to this change, merged all companion PRs, so that FTL + the init service have been adjusted long ago, why it makes Pi-hole future prove and the simple fact that currently the Docker image is pre-creating a directory that is never used and not creating a directory that is directly used by FTL, which only works as currently a symlink exists to maintain backwards compatibility, which will or will not stay in place, we don't know.

I know it fits well into your picture that all my PRs have no reason, but please at least try to gather related information before judging my commits.

**Comment by dschaper** [comment]
You asked for my review. I gave it to you.

The Pi-hole docker image has over 500 million pulls. The code we have now works. From your own words:

> It has not been tested

and 

> To be true, I am no Docker specialist. 

and

>which will or will not stay in place (on the underlying Docker base image), we don't know.

The image works, this change doesn't do anything to improve the image. `/run/` or `/var/run` is not a volume that is mounted. There is no reason for a user to `exec` in to a running container to have any interaction with those directories or files. 

This PR has been untouched for a year so I don't see a need to merge it in when no one has thought or talked about it for a year.

On a personal note, continuing to argue and post diatribes like you did in the core repo because you were told "no" is not helping your case. I don't trust your code anymore and the lack of ability to accept criticism or corrections shows me that my lack of trust has a valid basis.

**Comment by MichaIng** [comment]
> is not helping your case. I don't trust your code anymore and the lack of ability to accept criticism or corrections shows me that my lack of trust has a valid basis.

It's not my project you block reasonable and foreseeing development on, because of personal dislike despite, it seems. That's sad, but I'm accepting it and hence closing this PR.

**Pull request closed without merging**